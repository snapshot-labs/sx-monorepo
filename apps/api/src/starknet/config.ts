import { CheckpointConfig } from '@snapshot-labs/checkpoint';
import { starknetNetworks } from '@snapshot-labs/sx';
import { validateAndParseAddress } from 'starknet';
import { DEFAULT_INFURA_API_KEY } from '../config';
import spaceAbi from './abis/space.json';
import spaceFactoryAbi from './abis/spaceFactory.json';

const snNetworkNodeUrl =
  process.env.NETWORK_NODE_URL_SN ||
  `https://starknet-mainnet.infura.io/v3/${DEFAULT_INFURA_API_KEY}`;
const snSepNetworkNodeUrl =
  process.env.NETWORK_NODE_URL_SN_SEP ||
  `https://starknet-sepolia.infura.io/v3/${DEFAULT_INFURA_API_KEY}`;
const manaRpcUrl = process.env.VITE_MANA_URL || 'https://mana.box';

export type FullConfig = {
  indexerName: 'sn' | 'sn-sep';
  overrides: ReturnType<typeof createOverrides>;
} & CheckpointConfig;

const CONFIG = {
  sn: {
    indexerName: 'sn',
    networkNodeUrl: snNetworkNodeUrl,
    l1NetworkNodeUrl: 'https://rpc.brovider.xyz/1',
    contract: starknetNetworks['sn'].Meta.spaceFactory,
    start: 445498,
    verifiedSpaces: [
      '0x009fedaf0d7a480d21a27683b0965c0f8ded35b3f1cac39827a25a06a8a682a4',
      '0x05ea5ef0c54c84dc7382629684c6e536c0b06246b3b0981c426b42372e3ef263',
      '0x07c251045154318a2376a3bb65be47d3c90df1740d8e35c9b9d943aa3f240e50',
      '0x07bd3419669f9f0cc8f19e9e2457089cdd4804a4c41a5729ee9c7fd02ab8ab62'
    ]
  },
  'sn-sep': {
    indexerName: 'sn-sep',
    networkNodeUrl: snSepNetworkNodeUrl,
    l1NetworkNodeUrl: 'https://rpc.brovider.xyz/11155111',
    contract: starknetNetworks['sn-sep'].Meta.spaceFactory,
    start: 17960,
    verifiedSpaces: [
      '0x0141464688e48ae5b7c83045edb10ecc242ce0e1ad4ff44aca3402f7f47c1ab9'
    ]
  }
};

function createOverrides(networkId: keyof typeof CONFIG) {
  const config = starknetNetworks[networkId];

  return {
    networkNodeUrl: CONFIG[networkId].networkNodeUrl,
    l1NetworkNodeUrl: CONFIG[networkId].l1NetworkNodeUrl,
    manaRpcUrl: `${manaRpcUrl}/stark_rpc/${config.Meta.eip712ChainId}`,
    baseChainId: config.Meta.herodotusAccumulatesChainId,
    erc20VotesStrategy: config.Strategies.ERC20Votes,
    propositionPowerValidationStrategyAddress:
      config.ProposalValidations.VotingPower,
    spaceClassHash: config.Meta.masterSpace,
    verifiedSpaces: CONFIG[networkId].verifiedSpaces,
    herodotusStrategies: [
      config.Strategies.OZVotesStorageProof,
      config.Strategies.OZVotesTrace208StorageProof,
      config.Strategies.EVMSlotValue
    ]
      .filter(address => !!address)
      .map(strategy => validateAndParseAddress(strategy))
  };
}

export function createConfig(indexerName: keyof typeof CONFIG): FullConfig {
  const { networkNodeUrl, contract, start } = CONFIG[indexerName];

  const overrides = createOverrides(indexerName);

  return {
    indexerName,
    overrides,
    network_node_url: networkNodeUrl,
    optimistic_indexing: true,
    sources: [
      {
        contract,
        start,
        abi: 'SpaceFactory',
        events: [
          {
            name: 'NewContractDeployed',
            fn: 'handleContractDeployed'
          }
        ]
      }
    ],
    templates: {
      Space: {
        abi: 'Space',
        events: [
          {
            name: 'SpaceCreated',
            fn: 'handleSpaceCreated'
          },
          {
            name: 'ProposalCreated',
            fn: 'handlePropose'
          },
          {
            name: 'ProposalCancelled',
            fn: 'handleCancel'
          },
          {
            name: 'ProposalUpdated',
            fn: 'handleUpdate'
          },
          {
            name: 'ProposalExecuted',
            fn: 'handleExecute'
          },
          {
            name: 'VoteCast',
            fn: 'handleVote'
          },
          {
            name: 'MetadataUriUpdated',
            fn: 'handleMetadataUriUpdated'
          },
          {
            name: 'MinVotingDurationUpdated',
            fn: 'handleMinVotingDurationUpdated'
          },
          {
            name: 'MaxVotingDurationUpdated',
            fn: 'handleMaxVotingDurationUpdated'
          },
          {
            name: 'VotingDelayUpdated',
            fn: 'handleVotingDelayUpdated'
          },
          {
            name: 'OwnershipTransferred',
            fn: 'handleOwnershipTransferred'
          },
          {
            name: 'AuthenticatorsAdded',
            fn: 'handleAuthenticatorsAdded'
          },
          {
            name: 'AuthenticatorsRemoved',
            fn: 'handleAuthenticatorsRemoved'
          },
          {
            name: 'VotingStrategiesAdded',
            fn: 'handleVotingStrategiesAdded'
          },
          {
            name: 'VotingStrategiesRemoved',
            fn: 'handleVotingStrategiesRemoved'
          },
          {
            name: 'ProposalValidationStrategyUpdated',
            fn: 'handleProposalValidationStrategyUpdated'
          }
        ]
      }
    },
    abis: {
      SpaceFactory: spaceFactoryAbi,
      Space: spaceAbi
    }
  };
}
