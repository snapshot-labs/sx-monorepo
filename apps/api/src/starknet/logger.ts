import logger from '../logger';

export default logger.child({ module: 'starknet', protocol: 'snapshotX' });
