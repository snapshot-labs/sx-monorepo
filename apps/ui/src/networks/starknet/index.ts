import {
  LibraryError,
  ReceiptTx,
  constants as starknetConstants
} from 'starknet';
import { pinPineapple } from '@/helpers/pin';
import { Network } from '@/networks/types';
import { NetworkID, Space } from '@/types';
import { createActions } from './actions';
import { createConstants } from './constants';
import { createProvider } from './provider';
import { STARKNET_CONNECTORS } from '../common/constants';
import { createApi } from '../common/graphqlApi';

type Metadata = {
  name: string;
  chainId: starknetConstants.StarknetChainId;
  baseChainId: number;
  baseNetworkId: NetworkID;
  rpcUrl: string;
  ethRpcUrl: string;
  explorerUrl: string;
  apiUrl: string;
  avatar: string;
};

export const METADATA: Partial<Record<NetworkID, Metadata>> = {
  sn: {
    name: 'Starknet',
    chainId: starknetConstants.StarknetChainId.SN_MAIN,
    baseChainId: 1,
    baseNetworkId: 'eth',
    rpcUrl: `https://starknet-mainnet.infura.io/v3/${import.meta.env.VITE_INFURA_API_KEY}`,
    ethRpcUrl: `https://mainnet.infura.io/v3/${import.meta.env.VITE_INFURA_API_KEY}`,
    apiUrl: 'https://api.snapshot.box',
    explorerUrl: 'https://starkscan.co',
    avatar: 'ipfs://bafkreihbjafyh7eud7r6e5743esaamifcttsvbspfwcrfoc5ykodjdi67m'
  },
  'sn-sep': {
    name: 'Starknet Sepolia',
    chainId: starknetConstants.StarknetChainId.SN_SEPOLIA,
    baseChainId: 11155111,
    baseNetworkId: 'sep',
    rpcUrl: `https://starknet-sepolia.infura.io/v3/${import.meta.env.VITE_INFURA_API_KEY}`,
    ethRpcUrl: `https://sepolia.infura.io/v3/${import.meta.env.VITE_INFURA_API_KEY}`,
    apiUrl:
      import.meta.env.VITE_STARKNET_SEPOLIA_API ??
      'https://testnet-api.snapshot.box',
    explorerUrl: 'https://sepolia.starkscan.co',
    avatar: 'ipfs://bafkreihbjafyh7eud7r6e5743esaamifcttsvbspfwcrfoc5ykodjdi67m'
  }
};

export function createStarknetNetwork(networkId: NetworkID): Network {
  const metadata = METADATA[networkId];
  if (!metadata) throw new Error(`Unsupported network ${networkId}`);

  const {
    name,
    chainId,
    baseChainId,
    baseNetworkId,
    rpcUrl,
    ethRpcUrl,
    apiUrl,
    explorerUrl,
    avatar
  } = metadata;

  const provider = createProvider(rpcUrl);
  const constants = createConstants(networkId, baseNetworkId, baseChainId);
  const api = createApi(apiUrl, networkId, constants, {
    baseNetworkId
  });

  const helpers = {
    isAuthenticatorSupported: (authenticator: string) =>
      constants.SUPPORTED_AUTHENTICATORS[authenticator],
    isAuthenticatorContractSupported: (authenticator: string) =>
      constants.CONTRACT_SUPPORTED_AUTHENTICATORS[authenticator],
    getRelayerAuthenticatorType: (authenticator: string) =>
      constants.RELAYER_AUTHENTICATORS[authenticator],
    isStrategySupported: (strategy: string) =>
      constants.SUPPORTED_STRATEGIES[strategy],
    isExecutorSupported: (executor: string) =>
      constants.SUPPORTED_EXECUTORS[executor],
    pin: pinPineapple,
    getSpaceController: async (space: Space) => space.controller,
    getTransaction: txId => provider.getTransactionReceipt(txId),
    waitForTransaction: txId => {
      let retries = 0;

      return new Promise((resolve, reject) => {
        const timer = setInterval(async () => {
          let tx: Awaited<ReturnType<typeof provider.getTransactionReceipt>>;
          try {
            tx = await provider.getTransactionReceipt(txId);
          } catch (e) {
            if (
              e instanceof LibraryError &&
              e.message.includes('Transaction hash not found')
            ) {
              if (retries > 60) {
                clearInterval(timer);
                reject();
              }

              retries++;
            }

            return;
          }

          const receiptTx = new ReceiptTx(tx);
          if (receiptTx.isSuccess()) {
            resolve(tx);
          } else {
            reject(tx);
          }

          clearInterval(timer);
        }, 2000);
      });
    },
    waitForSpace: (spaceAddress: string, interval = 5000): Promise<Space> =>
      new Promise(resolve => {
        const timer = setInterval(async () => {
          const space = await api.loadSpace(spaceAddress);
          if (space) {
            clearInterval(timer);
            resolve(space);
          }
        }, interval);
      }),
    getExplorerUrl: (id, type) => {
      let dataType: 'tx' | 'contract' | 'token' = 'tx';
      if (type === 'token') dataType = 'token';
      else if (['address', 'contract', 'strategy'].includes(type))
        dataType = 'contract';

      return `${explorerUrl}/${dataType}/${id}`;
    }
  };

  return {
    name,
    avatar,
    currentUnit: 'second',
    chainId,
    baseChainId,
    currentChainId: baseChainId,
    baseNetworkId,
    supportsSimulation: true,
    managerConnectors: STARKNET_CONNECTORS,
    actions: createActions(networkId, provider, helpers, {
      chainId,
      l1ChainId: baseChainId,
      ethUrl: ethRpcUrl
    }),
    api,
    constants,
    helpers
  };
}
