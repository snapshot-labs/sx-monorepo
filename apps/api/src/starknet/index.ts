import Checkpoint, { starknet } from '@snapshot-labs/checkpoint';
import { createConfig } from './config';
import { createWriters } from './writers';

const snConfig = createConfig('sn');
const snSepConfig = createConfig('sn-sep');

const snIndexer = new starknet.StarknetIndexer(createWriters(snConfig));
const snSepIndexer = new starknet.StarknetIndexer(createWriters(snSepConfig));

export function addStarknetIndexers(checkpoint: Checkpoint) {
  checkpoint.addIndexer(snConfig.indexerName, snConfig, snIndexer);
  checkpoint.addIndexer(snSepConfig.indexerName, snSepConfig, snSepIndexer);
}
