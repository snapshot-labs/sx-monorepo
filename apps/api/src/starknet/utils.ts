import { getAddress } from '@ethersproject/address';
import { BigNumber } from '@ethersproject/bignumber';
import { Contract as EthContract } from '@ethersproject/contracts';
import { StaticJsonRpcProvider } from '@ethersproject/providers';
import { utils } from '@snapshot-labs/sx';
import {
  BigNumberish,
  CallData,
  Contract,
  RpcProvider,
  shortString,
  validateAndParseAddress
} from 'starknet';
import EncodersAbi from './abis/encoders.json';
import ExecutionStrategyAbi from './abis/executionStrategy.json';
import SimpleQuorumExecutionStrategyAbi from './abis/l1/SimpleQuorumExecutionStrategy.json';
import { FullConfig } from './config';
import { Space } from '../../.checkpoint/models';
import { handleVotingPowerValidationMetadata } from '../common/ipfs';
import logger from '../logger';

type StrategyConfig = {
  address: BigNumberish;
  params: BigNumberish[];
};

const encodersAbi = new CallData(EncodersAbi);

export function toAddress(bn: any) {
  try {
    return getAddress(BigNumber.from(bn).toHexString());
  } catch {
    return bn;
  }
}

export function longStringToText(array: string[]): string {
  return array.reduce(
    (acc, slice) => acc + shortString.decodeShortString(slice),
    ''
  );
}

export function findVariant(value: { variant: Record<string, any> }) {
  const result = Object.entries(value.variant).find(
    ([, v]) => typeof v !== 'undefined'
  );
  if (!result) throw new Error('Invalid variant');

  return {
    key: result[0],
    value: result[1]
  };
}

function formatAddress(type: string, address: string) {
  if (type === 'Starknet') {
    return validateAndParseAddress(address);
  }

  if (type === 'Ethereum') {
    const paddedAddress = `0x${address.replace('0x', '').padStart(40, '0')}`;
    return getAddress(paddedAddress);
  }

  return address;
}

export function formatAddressVariant({
  key,
  value
}: {
  key: string;
  value: string;
}) {
  return {
    type: key === 'Starknet' ? 0 : key === 'Ethereum' ? 1 : 2,
    address: formatAddress(key, value)
  };
}

export function getVoteValue(label: string) {
  if (label === 'Against') return 2;
  if (label === 'For') return 1;
  if (label === 'Abstain') return 3;

  throw new Error('Invalid vote label');
}

export async function handleExecutionStrategy(
  address: string,
  payload: string[],
  config: FullConfig
) {
  try {
    if (address === '0x0') return null;

    const starkProvider = new RpcProvider({
      nodeUrl: config.overrides.networkNodeUrl
    });

    const executionContract = new Contract(
      ExecutionStrategyAbi,
      address,
      starkProvider
    );

    const executionStrategyType = shortString.decodeShortString(
      await executionContract.get_strategy_type()
    );

    let quorum = 0n;
    let destinationAddress = null;
    if (executionStrategyType === 'SimpleQuorumVanilla') {
      quorum = await executionContract.quorum();
    } else if (executionStrategyType === 'EthRelayer') {
      const [l1Destination] = payload;
      if (!l1Destination)
        throw new Error('Invalid payload for EthRelayer execution strategy');
      destinationAddress = formatAddress('Ethereum', l1Destination);

      const ethProvider = new StaticJsonRpcProvider(
        config.overrides.l1NetworkNodeUrl,
        config.overrides.baseChainId
      );

      const SimpleQuorumExecutionStrategyContract = new EthContract(
        destinationAddress,
        SimpleQuorumExecutionStrategyAbi,
        ethProvider
      );

      quorum = (
        await SimpleQuorumExecutionStrategyContract.quorum()
      ).toBigInt();
    }

    return {
      executionStrategyType,
      destinationAddress,
      quorum
    };
  } catch (err) {
    logger.warn({ err }, 'Failed to get execution strategy type');

    return null;
  }
}

export async function updateProposalValidationStrategy(
  space: Space,
  validationStrategyAddress: string,
  validationStrategyParams: string[],
  metadataUri: string[],
  config: FullConfig
) {
  space.validation_strategy = validationStrategyAddress;
  space.validation_strategy_params = validationStrategyParams.join(',');
  space.voting_power_validation_strategy_strategies = [];
  space.voting_power_validation_strategy_strategies_params = [];
  space.voting_power_validation_strategy_metadata =
    longStringToText(metadataUri);

  if (
    utils.encoding.hexPadLeft(validationStrategyAddress) ===
    utils.encoding.hexPadLeft(
      config.overrides.propositionPowerValidationStrategyAddress
    )
  ) {
    const parsed = encodersAbi.parse(
      'proposition_power_params',
      validationStrategyParams
    ) as Record<string, any>;

    if (Object.keys(parsed).length !== 0) {
      space.proposal_threshold = parsed.proposal_threshold.toString(10);
      space.voting_power_validation_strategy_strategies =
        parsed.allowed_strategies.map(
          (strategy: StrategyConfig) => `0x${strategy.address.toString(16)}`
        );
      space.voting_power_validation_strategy_strategies_params =
        parsed.allowed_strategies.map((strategy: StrategyConfig) =>
          strategy.params.map(param => `0x${param.toString(16)}`).join(',')
        );
    }

    try {
      await handleVotingPowerValidationMetadata(
        space.id,
        space.voting_power_validation_strategy_metadata,
        config
      );
    } catch (err) {
      logger.warn({ err }, 'Failed to handle voting power strategies metadata');
    }
  }
}

export async function registerProposal(
  {
    l1TokenAddress,
    strategyAddress,
    snapshotTimestamp
  }: {
    l1TokenAddress: string;
    strategyAddress: string;
    snapshotTimestamp: number;
  },
  config: FullConfig
) {
  const res = await fetch(config.overrides.manaRpcUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      jsonrpc: '2.0',
      id: 0,
      method: 'registerProposal',
      params: {
        l1TokenAddress,
        strategyAddress,
        snapshotTimestamp
      }
    })
  });

  return res.json();
}
