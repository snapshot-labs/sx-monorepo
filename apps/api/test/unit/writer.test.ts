import { describe, it, expect } from 'vitest';
import { mockDeep } from 'vitest-mock-extended';
import { AsyncMySqlPool } from '@snapshot-labs/checkpoint';
import { validateAndParseAddress } from 'starknet';

async function handleDeploy({ source, block, tx, mysql }) {
  console.log('Handle deploy');
  const item = {
    id: validateAndParseAddress(source.contract),
    name: 'Pistachio DAO',
    voting_delay: 3600,
    voting_period: 86400,
    proposal_threshold: 1,
    proposal_count: 0,
    vote_count: 0,
    created: block.timestamp,
    tx: tx.transaction_hash
  };
  const query = `INSERT IGNORE INTO spaces SET ?;`;
  await mysql.queryAsync(query, [item]);
}

describe('handleDeploy', () => {
  it('should create space with correct parameters', async () => {
    // TODO: Replace inputs with typed mocks

    const source = {
      contract: '0x0625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f'
    };

    const block = {
      timestamp: 1652650963200
    };

    const tx = {
      transaction_hash: '0xabbb5161c49978ffc76369df76cdb40683e33c09145a9681553fe8e9c9a96a46'
    };

    const mockMysql = mockDeep<AsyncMySqlPool>();

    await handleDeploy({
      source,
      block,
      tx,
      mysql: mockMysql
    });

    expect(mockMysql.queryAsync.mock.calls).toMatchSnapshot();
  });
});
