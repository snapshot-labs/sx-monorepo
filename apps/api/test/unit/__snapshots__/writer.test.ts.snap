// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handleDeploy > should create space with correct parameters 1`] = `
[
  [
    "INSERT IGNORE INTO spaces SET ?;",
    [
      {
        "created": 1652650963200,
        "id": "0x00625dc1290b6e936be5f1a3e963cf629326b1f4dfd5a56738dea98e1ad31b7f",
        "name": "Pistachio DAO",
        "proposal_count": 0,
        "proposal_threshold": 1,
        "tx": "0xabbb5161c49978ffc76369df76cdb40683e33c09145a9681553fe8e9c9a96a46",
        "vote_count": 0,
        "voting_delay": 3600,
        "voting_period": 86400,
      },
    ],
  ],
]
`;
